src/Services/FHIR/FhirObservationService.php

```diff
@@ -0,0 +1,233 @@
+{
+    private $FHIRData;
+
+    public function __construct()
+    {
+        parent::__construct('form_vitals');
+    }
+
+    public function createObservationResource($id = '', $data = '', $encode = true)
+    {
+        $resource = new FHIRObservation();
+
+        // Check Profile type and add its Category | Status | Note
+        switch ($data['formdir']) {
+            case 'vitals':
+                $code = new FHIRCoding();
+                $coding = new FHIRCodeableConcept();
+                $code->setSystem("http://terminology.hl7.org/CodeSystem/observation-category");
+                $code->setCode("vital-signs");
+                $code->setDisplay("Vital Signs");
+                $coding->addCoding($code);
+                $coding->setText("Vital Signs");
+                $resource->addCategory($coding);
+                $data['text'] ? $resource->addNote($data['text']) : null;
+                $resource->setStatus('final');
+                break;
+            default:
+                #TODO: Return CapabilityStatement or Outcome Resource
+                break;
+        }
+
+        // Check Profile or Member and add/set appropriately
+        if ($data['profile'] == 'vitals') {
+            $this->addMembers($data, $resource);
+        } else {
+            $this->setValues($data, $resource);
+        }
+
+        $subject = new FHIRReference();
+        $subject->setReference("Patient/" . $data['pid']);
+        $subject->setType("Patient");
+        $resource->setSubject($subject);
+        $encounter =  new FHIRReference();
+        $encounter->setReference("Encounter/" . $data['encounter']);
+        $encounter->setType("Encounter");
+        $resource->setEncounter($encounter);
+        $performer = new FHIRReference();
+        $performer->setReference("Practitioner/" . $data['provider_id']);
+        $performer->setType("Practitioner");
+        $resource->addPerformer($performer);
+        $resource->setEffectiveDateTime($data['date']);
+        $resource->setId($id);
+
+        if ($encode) {
+            return json_encode($resource);
+        } else {
+            return $resource;
+        }
+    }
+
+    public function getVital($id)
+    {
+        return $this->get(
+            array(
+                "where" => "WHERE form_vitals.id = ?",
+                "data" => array($id),
+                "join" => "JOIN forms fo on form_vitals.id = fo.form_id",
+                "limit" => 1
+            )
+        );
+    }
+
+    private function addMembers($data, $resource)
+    {
+        $members = array_filter($data, function ($k) {
+            if ($k != 0 && $k != 0.0 && $k != 0.00 && $k != null && $k != '') {
+                return $k;
+            }
+        });
+
+        foreach ($members as $key => $value) {
+            if ($key == 'temperature') {
+                $temp_refrence = new FHIRReference();
+                $temp_refrence->setReference("Observation/temperature-" . $data['form_id']);
```

Comment:
we can't ref a form like this. needs to be Observation/body-temperature etc
This reference is saying subject for profile has a Observation/body-temperature member. Not a link to body-temperature.
Also, if not a profile then values become a component with codes etc. This is where the R4 classes become helpful.

Reply:
As its a Vitals-Panel so i think hasMember is necessary.
-> http://hl7.org/fhir/R4/vitalspanel.html
-> http://hl7.org/fhir/R4/observation-example-vitals-panel.json.html

Reply:
Did you mean this should be like 
-> Observation/body-temperature
or
-> Observation/body-temperature-id 😕

Reply:
Any reference to an observation component won't have an id because we don't store observations. we refer to them by the subject/codes/ect.

Of course for a profile they will have members where each member is an observation. The observation itself will contain the components when it gets called. It might make sense to write the canonical url using the search queries necessary to reach the observation!
dig a little deeper into the spec.

I'm still thinking that in our environment, here would be an ideal place for contained.
@rhausam do you have an opinion here?

Reply:
As we always start with getOne function in any Resource on 1st day i did start with that and that when i asked you that can we use `observation form` because I couldn't find any specific id for observation and we can't search it without the id. So try tried to search and walk around our DB and then I found two specific tables 1st `forms` , 2nd `forms_vitals` and after analysing the structure. what these forms table is storing can be considered as categories of Observetion Like :- Vital Sign, Exam , Laboratry etc. And the `form_vital` can be used as the specific Profile like in this case Vital-Sign. So That was the point when i decided to use `forms_vital.id` as Observation ID. AND as we are storing all Observation under one table all will have same id like - `body-temprature-5` will be picked from same `vital-5`. And this approach is working fine for me.

https://www.loom.com/share/97d8194c8c494da089d9a7a3dd5795a4
I tested it out and works. If you think it has any flaw or we shouldn't use it that I can definitely try something else or remove this.

Thanks :))

Reply:
I get what you're trying to do but we need to stay in bounds and try not to create custom searches. We need to not modify normal category/code. If you know you will parse vital panels by encounter vital form id then okay but, use Observation/{id} and leave the category/code the way the Fhir world knows about ie body-temperature.

Also for vitals remember search by date will be important because that would be a _history

Reply:
> use Observation/{id} and leave the category/code the way the Fhir world knows about ie body-temperature.

I don't have deep knowledge in FHIR all I have is from reading the spec.
1. Category codes don't have any code for the profiles http://hl7.org/fhir/R4/codesystem-observation-category.html#4.3.14.232.2.
2. And I don't think we are bound to use any code Like they show here -> http://hl7.org/fhir/R4/observation-example-vitals-panel.json.html.
because its just one example and when i go through few test servers like  -> http://wildfhir4.aegis.net/fhir4-0-0-gui/index.jsf they are using any reference like we always do. (Example  - Observation/{id} , Patient/{id})
![image](https://user-images.githubusercontent.com/47683556/79275141-4a63d300-7ec3-11ea-9f02-784fb883aae3.png)


I think @rhausam ser has lot more knowledge than me so better if they give there opinion

Reply:
Yes but Observations do! 
check
https://www.hl7.org/fhir/observation-example-vitals-panel.json.html
and
referencing the vitals form for a vitals profile(making the vitals form and observation object for the profile) is one thing but the individual vitals component category is another.
Besides all the components of the vitals profile will end up with the same id reference so it's really not a reference to body temperature but to the profile(vitals form in this case) instead.
Also depends on what value set you are using. Other servers may have their own defining how a category is named.
We're not a server. We are not storing Observation components that would have an id. All our reference to Observation components will be a search one. I don't want prefixed reference.

I don't know how to say differently. Remember vitals profile is not the only profile. We need to be consistent.

Reply:
> referencing the vitals form for a vitals profile(making the vitals form an observation object for the profile) is one thing but the individual vitals component category is another.

I totally agree with this, So should just keep Vital Panel with All members as components? 

BTW, I just finished the getBundle() and want to show [this.](https://github.com/yashrajbothra/openemr/blob/master/example.json) Although as we changed our mind I cant push this now :(

Reply:
I'm beginning to think we don't need this profile as all it does is tell client what vital profiles are available for a given patient. As for having the members reference to a vital profile say, body temperature to a form id, i'm uncertain this is appropriate but leave it as you have for now.

We really need to ensure search is done. Check this out for search:
https://www.hl7.org/fhir/observation-vitalsigns.html#Quick_Start

You're bundle doesn't look right but i'm super tired.

Reply:
I think it's great that you are making progress.  But I'm working on a (hopefully near) final push on another project and I have only been able to monitor peripherally in the last 48 hours.  I'll try to review and add some more substantial comments a bit later today.